firedrake/fml/form_manipulation_language.py

```diff
@@ -0,0 +1,484 @@
+"""A language for manipulating forms using labels."""
+
+import ufl
+import functools
+import operator
+from firedrake import Constant, Function
+
+
+__all__ = ["Label", "Term", "LabelledForm", "identity", "drop", "all_terms",
+           "keep", "subject", "name_label"]
+
+# ---------------------------------------------------------------------------- #
+# Core routines for filtering terms
+# ---------------------------------------------------------------------------- #
+identity = lambda t: t
+drop = lambda t: None
+all_terms = lambda t: True
+keep = identity
+
+
+# ---------------------------------------------------------------------------- #
+# Term class
+# ---------------------------------------------------------------------------- #
+class Term(object):
+    """A Term object contains a form and its labels."""
+
+    __slots__ = ["form", "labels"]
+
+    def __init__(self, form, label_dict=None):
+        """
+        Parameters
+        ----------
+        form : :class:`ufl.Form`
```

Comment:
This example is fine if I remove `:class:` and the backticks, but immediately below:
- For `label_dict : dict, optional` if I do the same I get the error:
```
/home/jack/Documents/firedrake/firedrake/src/firedrake/firedrake/fml/form_manipulation_language.py:docstring of firedrake.fml.form_manipulation_language.Term:1:py:class reference target not found: optional
```
Optional is correct according to [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#parameters).
- For `label_dict : <backtick>dict<backtick>, optional` the error goes away, but the auto-generated docs have no links to the docs for `dict`.
- Only with `label_dict : :obj:<backtick>dict<backtick>, optional` do I get what I thought was the correct behaviour with no error and a link to [this part of the Python docs](https://docs.python.org/3/library/stdtypes.html#dict).

Similar things happen elsewhere in the docstrings, where objects or functions are referenced and no link is generated. Am I just doing something wrong?

At least when I included the RST roles the right links were generated in the html documentation. But the latest commit has all of the roles removed.

Reply:
@ReubenHill I agree with your suggestions, but did you read the above comment?

Reply:
No I did not. That is very strange! Does the most recent commit give you links everywhere it ought to?

Reply:
No, the backticks just prevent the error from occurring. The only way I know how to get the link is to add the RST role tag back in. I am not a numpydoc/RST/sphinx expert, but I do find it difficult to find concrete examples/tutorials online, especially ones that cover these sorts of edge cases.

Reply:
Well if it produces the correct documentation (with links), then surely that's what you need to do. Do you get links when you don't use backticks for our internal classes?

Does this imply our documentation compiler just not properly deal with numpy-style docstrings?

Reply:
When we did the petsc4py documentation we omitted the type part of the numpy docstring and used type hints instead ([example](https://gitlab.com/petsc/petsc/-/blob/main/src/binding/petsc4py/src/petsc4py/PETSc/Vec.pyx?ref_type=heads#L330)). Since I expect us to start using type hints at some point is it worth even doing this?

Reply:
This appears to be under discussion in numpydoc: https://github.com/numpy/numpydoc/issues/196

Reply:
Apparently the "correct" fix for this is to (as Connor has already mentioned) add type hinting, which I have now done. However, it _still_ doesn't solve the issue! It's better, but the documentation for `LabelledForm.lable_map()` which has arguments that are callables looks like:

> label_map(term_filter: ~collections.abc.Callable, map_if_true: ~collections.abc.Callable = <function <lambda>>, map_if_false: ~collections.abc.Callable = <function <lambda>>) → [LabelledForm]

Which is clearly wrong! Whatsmore I've now broken backward compatibility with Python3.8, so now I need to go and change a bunch of the type hints I've provided.

I'm wondering whether this is just a waste of time, it certainly feels like it. :slightly_frowning_face:

Reply:
I vote you leave in the RST if that provides the correct links

Reply:
Oh, it gets even better! To allow the _code to execute_ in Python 3.8 we need to import from the `typing`module, which gives us a bunch of [links to deprecated functionality](https://docs.python.org/3/library/typing.html#typing.Callable) in our documentation.

Reply:
Fixed as best I can, html docs for `label_map` still look terrible, but using functions rather than lambdas makes things at least parsable.